game-core/src/main/java/games/strategy/triplea/ui/DicePanel.java

```diff
@@ -27,59 +27,80 @@
     this.uiContext = uiContext;
     this.data = data;
-    setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));
+    setLayout(new GridBagLayout());
+    setBorder(BorderFactory.createEmptyBorder(8, 16, 0, 0));
   }
 
   void setDiceRollForBombing(final List<Die> dice, final int cost) {
     removeAll();
+    add(new JLabel("Cost: " + cost, SwingConstants.LEFT));
     add(create(dice));
-    add(Box.createVerticalGlue());
-    add(new JLabel("Cost:" + cost));
-    invalidate();
+    addBottomLabel(new JLabel());
   }
 
   /** Sets the dice roll to display. */
   public void setDiceRoll(final DiceRoll diceRoll) {
-    SwingAction.invokeNowOrLater(
-        () -> {
-          removeAll();
-          for (int i = 1; i <= data.getDiceSides(); i++) {
-            final List<Die> dice = diceRoll.getRolls(i);
-            if (dice.isEmpty()) {
-              continue;
-            }
-            add(new JLabel("Rolled at " + i + ":"));
-            add(create(diceRoll.getRolls(i)));
-          }
-          add(Box.createVerticalGlue());
-          add(new JLabel("Total hits:" + diceRoll.getHits()));
-          validate();
-          invalidate();
-          repaint();
-        });
+    removeAll();
+    final String hitsString = colorizeHitString(diceRoll.getHits());
+    add(
+        new JLabel(
+            "<html><b><font style='font-size:120%'>Total hits: " + hitsString,
+            SwingConstants.LEFT));
+    add(new JSeparator());
+
+    for (int i = 1; i <= data.getDiceSides(); i++) {
+      final List<Die> dice = diceRoll.getRolls(i);
+      if (dice.isEmpty()) {
+        continue;
+      }
+      add(makeDiceRolledLabel(dice, i));
+      add(create(dice));
+    }
+    addBottomLabel(new JLabel());
+    invalidate();
+    validate();
+    repaint();
+  }
+
+  private JLabel makeDiceRolledLabel(final List<Die> dice, final int value) {
+    final long hits = dice.stream().map(Die::getType).filter(Die.DieType.HIT::equals).count();
+    final String countString = dice.size() == 1 ? "1 die" : dice.size() + " dice";
+    final String hitsString = colorizeHitString(hits == 1 ? "1 hit" : hits + " hits");
+    return new JLabel("<html><b>Rolled " + countString + " at " + value + " (" + hitsString + "):");
+  }
+
+  private static String colorizeHitString(final Object hitsString) {
+    // On a dark theme, use red. Use a darker red with a light theme.
+    final String color = LookAndFeel.isCurrentLookAndFeelDark() ? "red" : "#8B0000";
+    return "<font color='" + color + "'>" + hitsString + "</font>";
+  }
+
+  private void add(final JComponent component) {
+    final GridBagConstraints constraints = new GridBagConstraints();
+    constraints.fill = GridBagConstraints.HORIZONTAL;
+    constraints.weightx = 1;
+    constraints.gridx = 0;
+    add(component, constraints);
+  }
+
+  private void addBottomLabel(final JLabel label) {
+    final GridBagConstraints constraints = new GridBagConstraints();
```

Comment:
FWIW, there is a type-safe builder for GridBagConstraints that can be used.

Is `gridy` default to `0`, if so, do we need to set `gridx` to `0` as well on line 91?

Reply:
No, the default value is "RELATIVE", not 0.

Reply:
Looks like GridBagConstraintsBuilder doesn't support setting gridy to GridBagConstraints.RELATIVE. So let's stick with this. A future PR can always expand the functionality of GridBagConstraintsBuilder to support this use case and refactor this code, if someone wants to contribute (from my perspective, I'm not sure it would add much since the current code isn't very convoluted - and might be more confusing with the extra abstraction).

Reply:
Ah, wasn't aware of the limitation. I'm not sure I'd call the type-safe builders necessarily abstraction; regardless, I do agree the current code is pretty straight forward.